lib/spack/spack/environment/environment.py

```diff
@@ -90,15 +90,24 @@ def env_root_path() -> str:
 def environment_name(path: Union[str, pathlib.Path]) -> str:
     """Human-readable representation of the environment.
 
-    This is the path for directory environments, and just the name
+    This is the path for independent environments, and just the name
     for managed environments.
     """
     env_root = pathlib.Path(env_root_path()).resolve()
-    path_str = str(path)
-    if path_str.startswith(str(env_root)):
-        return str(pathlib.Path(path_str).relative_to(env_root))
+    path_path = pathlib.Path(path)
+
+    # If an environment name / directory in the configured environment root resolves
+    # to the same location as the path of a given environment, we know the environment
+    # is tracked or managed else we return the path like all other independent
+    # environments.
+    #
+    # Additionally we must resolve the env_root / name path as the root might itself
+    # be a symlink to another location. (Common use case is ~/.spack/environments/
+    # pointing to a set of dotfiles managed in a git repository.)
```

Comment:
```suggestion
    # For a managed environment created in Spack, env.path is ENV_ROOT/NAME
    # For a managed environment from `spack env create`, the path is symlinked to ENV_ROOT/NAME
    # So if ENV_ROOT/NAME resolves to env.path we know the environment is tracked/managed.
    # Otherwise, it is an independent environment and  we return the path.
    #
    # We resolve both paths fully because the env_root itself could also be a symlink,
    # and any directory in env.path could be a symlink.
```